libs/ts-rest/core/src/lib/dsl.ts

```diff
@@ -29,25 +36,80 @@ export type AppRouteMutation = {
   contentType?: 'application/json' | 'multipart/form-data';
   body: unknown;
   query?: unknown;
+  headers?: unknown;
   summary?: string;
   description?: string;
   deprecated?: boolean;
   responses: Record<number, unknown>;
 };
 
+type ValidatedHeaders<
+  T extends AppRoute,
+  TOptions extends RouterOptions,
+  TOptionsApplied = ApplyOptions<T, TOptions>
+> = 'headers' extends keyof TOptionsApplied
+  ? TOptionsApplied['headers'] extends MixedZodError<infer A, infer B>
+    ? {
+        _error: 'Cannot mix plain object types with Zod objects for headers';
+        a: A;
+        b: B;
+      }
+    : T
+  : T;
+
 /**
  * Recursively process a router, allowing for you to define nested routers.
  *
  * The main purpose of this is to convert all path strings into string constants so we can infer the path
  */
-type RecursivelyProcessAppRouter<T extends AppRouter> = {
+type RecursivelyProcessAppRouter<
+  T extends AppRouter,
+  TOptions extends RouterOptions
+> = {
   [K in keyof T]: T[K] extends AppRoute
-    ? T[K]
+    ? ValidatedHeaders<T[K], TOptions>
     : T[K] extends AppRouter
-    ? RecursivelyProcessAppRouter<T[K]>
+    ? RecursivelyProcessAppRouter<T[K], TOptions>
     : T[K];
 };
 
+type RecursivelyApplyOptions<
+  TRouter extends AppRouter,
+  TOptions extends RouterOptions
+> = {
+  [TRouterKey in keyof TRouter]: TRouter[TRouterKey] extends AppRoute
+    ? Prettify<ApplyOptions<TRouter[TRouterKey], TOptions>>
+    : TRouter[TRouterKey] extends AppRouter
+    ? RecursivelyApplyOptions<TRouter[TRouterKey], TOptions>
+    : TRouter[TRouterKey];
+};
+
+type UniversalMerge<A, B> = A extends z.AnyZodObject
```

Comment:
This literally blew my mind. Awesome. 

Need to spend a few days reviewing the whole PR to make sure I'm understanding everything fully! 😅

Reply:
Yeah that gave me a bit of a headache 😅 but I think it was worth not having the headache of trying to figure out how to combine Zod objects with plain TS objects.